File: web/js/entry.js
/** Show-host entry builder. */

/**
 * Tidy raw PokéAPI flavor text for on-screen display: old-game copy SHOUTS
 * "POKéMON", and the source often carries stray line breaks. Kept in step with
 * poke/entry.py:_clean_flavor. (Speech normalization is separate — tts_text.py.)
 * @param {string} text
 * @returns {string}
 */
function cleanFlavor(text) {
  return (text || "")
    .replace(/POKéMON/g, "Pokémon")
    .replace(/POKÉMON/g, "Pokémon")
    .replace(/\s+/g, " ")
    .trim();
}

/**
 * PokéAPI gender_rate → chip label. Mirrors poke/entry.py:gender_label.
 * @param {number} rate
 * @returns {string}
 */
function genderLabel(rate) {
  if (rate < 0) return "—";
  if (rate === 0) return "♂";
  if (rate === 8) return "♀";
  return "♂ ♀";
}

/**
 * @param {PokemonRecord} data
 * @returns {DexEntryView}
 */
function buildEntry(data) {
  var typesLine = data.types && data.types.length ? data.types.join(" / ") : "Unknown";
  var hw =
    data.heightM.toFixed(1) + " m · " + data.weightKg.toFixed(1) + " kg";
  var dexBit = data.dexNumber ? "National No. " + data.dexNumber + ". " : "";
  var abilities =
    data.abilities && data.abilities.length ? data.abilities.slice() : [];
  var weaknesses =
    data.weaknesses && data.weaknesses.length ? data.weaknesses.slice() : [];
  var stats = data.baseStats || {
    hp: 0,
    attack: 0,
    defense: 0,
    specialAttack: 0,
    specialDefense: 0,
    speed: 0,
  };
  var chain =
    data.evolutionChain && data.evolutionChain.length
      ? data.evolutionChain.map(function (s) {
          return { slug: s.slug, displayName: s.displayName };
        })
      : [{ slug: data.name, displayName: data.displayName }];

  var narration =
    data.displayName +
    "! " +
    dexBit +
    "The " +
    data.category +
    ". This Pokémon is " +
    typesLine +
    " type, standing " +
    data.heightM.toFixed(1) +
    " meters tall and weighing " +
    data.weightKg.toFixed(1) +
    " kilograms. " +
    data.flavorText +
    " " +
    data.evolutionNote;

  if (narration.length > 520) {
    narration = narration.slice(0, 500).replace(/\s+\S*$/, "") + "…";
  }

  return {
    title: data.displayName,
    dexNumber: data.dexNumber,
    typesLine: typesLine,
    category: data.category,
    heightWeight: hw,
    genderLabel: genderLabel(typeof data.genderRate === "number" ? data.genderRate : -1),
    abilities: abilities,
    weaknesses: weaknesses,
    baseStats: {
      hp: stats.hp || 0,
      attack: stats.attack || 0,
      defense: stats.defense || 0,
      specialAttack: stats.specialAttack || 0,
      specialDefense: stats.specialDefense || 0,
      speed: stats.speed || 0,
    },
    evolutionChain: chain,
    narration: narration.trim(),
    description: cleanFlavor(data.flavorText) || data.evolutionNote,
    attribution: "Data: PokéAPI snapshot (offline). Fan demo only.",
  };
}

window.PokeEntry = { buildEntry: buildEntry };

File: poke/entry.py
"""Build show-host-style Pokédex narration and structured entry fields."""

from __future__ import annotations

import re
from dataclasses import dataclass

from poke.api_client import BaseStats, EvolutionStage, PokemonData

# Old-game flavor text SHOUTS the species noun; render it as the proper word.
_SHOUT_POKEMON = re.compile(r"POK[éÉ]MON")


def _clean_flavor(text: str) -> str:
    """Tidy raw flavor text for on-screen display. Mirrors web/js/entry.js:cleanFlavor.

    Speech normalization is a separate, richer pass — see poke/tts_text.py.
    """
    return re.sub(r"\s+", " ", _SHOUT_POKEMON.sub("Pokémon", text or "")).strip()


def gender_label(rate: int) -> str:
    """PokéAPI gender_rate → chip label. Mirrors web/js/entry.js:genderLabel."""
    if rate < 0:
        return "—"
    if rate == 0:
        return "♂"
    if rate == 8:
        return "♀"
    return "♂ ♀"


@dataclass(frozen=True)
class DexEntry:
    title: str
    dex_number: int | None
    types_line: str
    category: str
    height_weight: str
    gender_label: str
    abilities: tuple[str, ...]
    weaknesses: tuple[str, ...]
    base_stats: BaseStats
    evolution_chain: tuple[EvolutionStage, ...]
    narration: str
    description: str
    facts: tuple[str, ...]
    attribution: str = "Data: PokéAPI snapshot (offline). Fan demo only."


def build_entry(data: PokemonData) -> DexEntry:
    types_line = " / ".join(data.types) if data.types else "Unknown"
    hw = f"{data.height_m:.1f} m · {data.weight_kg:.1f} kg"
    dex_bit = f"National No. {data.dex_number}. " if data.dex_number else ""
    abilities = data.abilities if data.abilities else ()
    g_label = gender_label(data.gender_rate)

    narration = (
        f"{data.display_name}! {dex_bit}"
        f"The {data.category}. "
        f"This Pokémon is {types_line} type, standing {data.height_m:.1f} meters tall "
        f"and weighing {data.weight_kg:.1f} kilograms. "
        f"{data.flavor_text} "
        f"{data.evolution_note}"
    )
    # Keep TTS in a comfortable 15–40s band for typical voices (estimate).
    if len(narration) > 520:
        narration = narration[:500].rsplit(" ", 1)[0] + "…"

    facts = (
        f"Type: {types_line}",
        f"Category: {data.category}",
        f"Height / Weight: {hw}",
        f"Gender: {g_label}",
        f"Ability: {', '.join(abilities) if abilities else 'Unknown'}",
        data.evolution_note,
    )

    return DexEntry(
        title=data.display_name,
        dex_number=data.dex_number,
        types_line=types_line,
        category=data.category,
        height_weight=hw,
        gender_label=g_label,
        abilities=abilities,
        weaknesses=data.weaknesses,
        base_stats=data.base_stats,
        evolution_chain=data.evolution_chain,
        narration=narration.strip(),
        description=_clean_flavor(data.flavor_text) or data.evolution_note,
        facts=facts,
    )
